compiler/src/iree/compiler/Dialect/Stream/IR/StreamTypes.cpp

```diff
@@ -67,7 +67,7 @@ static llvm::cl::opt<bool> clResourceAliasMutableBindings(
     "iree-stream-resource-alias-mutable-bindings",
     llvm::cl::desc(
         "Fuses bindings that are mutable instead of leaving them split."),
-    llvm::cl::init(false));
+    llvm::cl::init(true));
```

Comment:
Digging into why CUDA is so much slower on some programs with this flag.

I'm looking at `EfficientNetB7PT`:
* Download with `gcloud storage cp "gs://iree-github-actions-presubmit-artifacts/5859077781/2/e2e-test-artifacts/model_EfficientNetB7PT.mlir" ..\iree-tmp\2023_08_14\`
* Compile with `..\iree-build\tools\iree-compile.exe ..\iree-tmp\2023_08_14\model_EfficientNetB7PT.mlir --iree-hal-target-backends=cuda --iree-hal-cuda-llvm-target-arch=sm_60 -o ..\iree-tmp\2023_08_14\model_EfficientNetB7PT_baseline.vmfb` (using `sm_60` for my 2080 TI, a100s would use `sm_80`)
* Run with `..\iree-build\tools\iree-run-module.exe --module=..\iree-tmp\2023_08_14\model_EfficientNetB7PT_baseline.vmfb --device=cuda --function=forward --input=1x3x600x600xf32`
* Compare baseline to compiling with `--iree-stream-resource-alias-mutable-bindings=true` also set

Baseline (with the flag off, do not alias mutable bindings) has `iree_hal_executable_cache_prepare_executable` take 46ms total over 150 calls (300 us mean).
With the flag, `iree_hal_executable_cache_prepare_executable` takes 11s total over 150 calls (75ms mean).

Seems like most of that time is in nvcuda64.dll:
![image](https://github.com/openxla/iree/assets/4010439/5a9552ac-4e39-4c99-9180-c71375fb1b4b)

The IR changes after `IR Dump After FuseDispatchBindings (iree-stream-fuse-dispatch-bindings)` are simple and expected:  ![image](https://github.com/openxla/iree/assets/4010439/57df508c-fbb3-448d-86b0-2c64bf80cf66)
I guess I can look at the HAL executable IR next to see what could be so much slower in the GPU driver.

Reply:
HAL IR looks similar. Here's one example, where the first binding is changed from readonly, with a pretty large offset into the portion that is written into: 
![image](https://github.com/openxla/iree/assets/4010439/7c91da46-b139-4472-a7de-886f9fb90b05)

Reply:
Here's a trace comparison of executable creation, comparing between baseline and with aliasing: 
![image](https://github.com/openxla/iree/assets/4010439/a3572390-3e9d-492c-8329-098417fc5eee)

`iree_vm_invoke` also jumps from 163ms to 292ms

Reply:
Looks like executable creation time could be a red herring (I think my driver is caching something?). I also tested with `cuda2` (compiling with `--iree-stream-resource-alias-mutable-bindings=true --iree-stream-emulate-memset=true`). Next step could be to collect data with benchmarks (multiple iterations, warmup runs) instead of a single `iree-run-module` invocation.